pkg/cmd/cli/cmd/rsync.go

```diff
@@ -0,0 +1,416 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+	"gopkg.in/pipe.v2"
+	kclient "k8s.io/kubernetes/pkg/client"
+	kubecmd "k8s.io/kubernetes/pkg/kubectl/cmd"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+const (
+	// RsyncRecommendedName is the recommended name for the rsync command
+	RsyncRecommendedName = "rsync"
+
+	rsyncLong = `
+Copy local files to a container
+
+This command will attempt to copy local files to a remote container. It will default to the
+first container if none is specified, and will attempt to use 'rsync' if available locally and in the
+container. If 'rsync' is not present, it will attempt to use 'tar' to send files to the container.`
+
+	rsyncExample = `
+  // Syncrhonize a local directory with a pod directory
+  $ %[1]s POD ./local/dir/ /remote/dir`
+
+	defaultRsyncExecutable = "rsync"
+	defaultTarExecutable   = "tar"
+)
+
+var (
+	testRsyncCommand = []string{"rsync", "--version"}
+	testTarCommand   = []string{"tar", "--version"}
+)
+
+// RsyncOptions holds the options to execute the sync command
+type RsyncOptions struct {
+	Namespace       string
+	PodName         string
+	ContainerName   string
+	Source          string
+	Destination     string
+	RsyncExecutable string
+	TarExecutable   string
+	RshCommand      string
+	Quiet           bool
+
+	Out io.Writer
+	Err io.Writer
+
+	LocalExecutor  executor
+	RemoteExecutor executor
+	PipeExecutor   pipeExecutor
+}
+
+// executor executes commands
+type executor interface {
+	Execute(command []string, in io.Reader, out, err io.Writer) error
+}
+
+// pipeExecutor pipes the stdout/stdin streams of a sequence of commands
+type pipeExecutor interface {
+	Execute(out, err io.Writer, pipes ...pipe.Pipe) error
+}
+
+// NewCmdRsync creates a new sync command
+func NewCmdRsync(name, fullName string, f *clientcmd.Factory, out, errOut io.Writer) *cobra.Command {
+	o := RsyncOptions{
+		Out:           out,
+		Err:           errOut,
+		LocalExecutor: &defaultLocalExecutor{},
+		PipeExecutor:  &defaultPipeExecutor{},
+	}
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s POD LOCAL_SOURCE REMOTE_DESTINATION", name),
+		Short:   "Copy local files to a pod",
+		Long:    rsyncLong,
+		Example: fmt.Sprintf(rsyncExample, fullName),
+		Run: func(c *cobra.Command, args []string) {
+			kcmdutil.CheckErr(o.Complete(f, c, args))
+			kcmdutil.CheckErr(o.Validate())
+			kcmdutil.CheckErr(o.Run())
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", "", "Container within the pod")
+	cmd.Flags().StringVar(&o.RsyncExecutable, "rsync", "", "Path to rsync executable")
+	cmd.Flags().StringVar(&o.TarExecutable, "tar", "", "Path to tar executable")
```

Comment:
I would only need to look for rsync since we can do tar on the client. 
There's no rsync that works yet on windows without an additional binary to handle the stream redirection:
http://superuser.com/questions/277859/is-there-a-way-to-use-plink-with-rsync-on-windows